CR-07: Implement currency conversion logic

tartarJR opened the pull request.

flyingBear66 commented on app/src/main/java/com/tatar/currencify/data/feature/LatestRatesDataModel.kt:
```diff
@@ -0,0 +1,32 @@
+
+data class LatestRatesData(
+    val baseCurrency: String,
+    val ratesData: List<RateData>
```
I'd leave as rates or rateDataList. Because when think of something as data, I don't imagine as a list. But your call...

tartarJR replied:
Will refactor

flyingBear66 commented on app/src/main/java/com/tatar/currencify/data/feature/LatestRatesDataModel.kt:
```diff
@@ -0,0 +1,32 @@
+}
+
+data class RateData(
+    val currencyCode: String,
```
would it be better if you make currency codes `enum`  instead of `String`?

tartarJR replied:
It creates extra boilerplate since I use a different data class for each layer.

flyingBear66 commented on app/src/main/java/com/tatar/currencify/presentation/feature/CurrencyConversionViewModel.kt:
```diff
@@ -0,0 +1,120 @@
+    private val _responderClickEvent = MutableLiveData<Event<Unit>>()
+    val responderClickEvent: LiveData<Event<Unit>> get() = _responderClickEvent
+
+    private val _animationEvent = MutableLiveData<Event<Unit>>()
```
this event is a bit vague by it's name, like what does it do with animation. hideKeyboard Event for example, is pretty clear, but I'm having a hard time to understand what animation Event does by looking at the code. Could you make it more clear maybe?

tartarJR replied:
Will refactor

flyingBear66 commented on app/src/main/java/com/tatar/currencify/presentation/feature/CurrencyConversionViewModel.kt:
```diff
@@ -0,0 +1,120 @@
+        _currencyConversionResult.postValue(currencyConversionResult)
+    }
+
+    private fun isNewFlow(isLoading: Boolean?) = isLoading == true
```
`Boolean?` means what? Does it nullable?

tartarJR replied:
Yup

flyingBear66 replied:
isLoading can be nil you mean, how? It should be either loading or notLoading.

tartarJR replied:
`isLoading` value comes from the `_loadingState` member variable and `_loadingState` is a `LiveData` field. You can think of it some kind of an observable generic data holder. Initially `LiveData` values are set to null.

flyingBear66 commented on app/src/main/java/com/tatar/currencify/ui/feature/ConvertedCurrenciesAdapter.kt:
```diff
@@ -0,0 +1,137 @@
+}
+
+class ConvertedCurrenciesDiffUtil(
+    private val oldList: List<ConvertedCurrency>,
```
Aren't `ConvertedCurrency` list always have unique elements? Maybe you can use `Set` and you comparisons would be better. E.g. in `areItemsTheSame`. Or for you the order is important somehow?

tartarJR replied:
This is an Android specific thing. `ConvertedCurrenciesDiffUtil` is used to optimise the performance of the `Recyclerview`. The point is comparing the old and new list items and change only the necessary fields in the `Recyclerview`.

flyingBear66 commented on app/src/main/java/com/tatar/currencify/ui/feature/CurrencyConversionActivity.kt:
```diff
@@ -0,0 +1,234 @@
+        currencyConversionViewModel.loadingState.observe(
+            this,
+            Observer { isLoading ->
+                if (isLoading) {
```
I'd prefer `toggle` methods instead of `hide` and `show` methods. This way you could have less method and your Observer function would be shorter. I know that's personal preference but I would make my helper functions more crowded than my observer,

flyingBear66 replied:
And to show how I would do it 
```
    private fun toggleLoadingIndicator(show: boolean, binding: ActivityCurrencyConversionBinding) {
        binding.progressBar.visibility = show ? View.VISIBLE : View.GONE
    }
```
instead of 
```
    private fun displayLoadingIndicator(binding: ActivityCurrencyConversionBinding) {
        binding.progressBar.visibility = View.VISIBLE
    }

    private fun hideLoadingIndicator(binding: ActivityCurrencyConversionBinding) {
        binding.progressBar.visibility = View.GONE
    }
```

tartarJR replied:
I sometimes do that too but I think it decreases the readability of the code. One decreases readability and other creates more boilerplate. I would prefer the better readability.

flyingBear66 replied:
🤷‍♂️

flyingBear66 commented on app/src/main/java/com/tatar/currencify/ui/util/UIExtensions.kt:
```diff
@@ -0,0 +1,113 @@
+
+fun String.toCurrencyString(): String {
+    return when (this) {
+        CURRENCY_CODE_AUD -> "Australian Dollar"
```
Woow, that's something. I think my enum suggestion would work on this kind of case, what you think?

flyingBear66 replied:
instead of creating every code const string.

tartarJR replied:
Enums can be used instead of the String but this time I have to create enum type  for each currency. Does not really add a value plus more boilerplate issue that I mentioned above.

flyingBear66 reviewed (commented):
Have some questions and suggestions but overall pretty nice structure 👍 . Let's talk about them.

flyingBear66 reviewed (approved):
LGTM!

tartarJR committed:
CR-07: Implement currency conversion logic

CR-07: Minor UI improvements

CR-07: Add currency conversion logic implementation

CR-07: Refactor class/field names to make more sense

CR-07: Refactor based on the code review
